src/librustc_codegen_ssa/back/write.rs

```diff
@@ -147,17 +150,6 @@ impl ModuleConfig {
         self.no_builtins = no_builtins || sess.target.target.options.no_builtins;
         self.inline_threshold = sess.opts.cg.inline_threshold;
         self.new_llvm_pass_manager = sess.opts.debugging_opts.new_llvm_pass_manager;
-        self.obj_is_bitcode =
-            sess.target.target.options.obj_is_bitcode || sess.opts.cg.linker_plugin_lto.enabled();
-        self.embed_bitcode =
-            if sess.target.target.options.embed_bitcode || sess.opts.debugging_opts.embed_bitcode {
-                match sess.opts.optimize {
-                    config::OptLevel::No | config::OptLevel::Less => EmbedBitcode::Marker,
```

Comment:
Hm ok so after some digging, here's some information about the "bitcode marker" situation.

I dug up [these questions](https://github.com/rust-lang/rust/issues/35968#issuecomment-371558729). Remember though that this all seems related to iOS circa 3 years ago, things may and probably have changed in the interim. It sounds like the linker complains if *anything* looks like it doesn't have bitcode in it, so the goal here was to avoid having Rust code causing the linker to complain. Or at least that's my best read.

I think in light of that we may want to preserve this in some form. Could we perhaps have something like:

```rust
enum EmitObj {
    Object { bitcode_marker: bool }
}
```

where that defaults to `false`? That way we can select this with logic that looks like:

* If bitcode embedding is requested, we'll be using `EmitObj::ObjectPlusBitcode`
* If no bitcode embedding is requested, then the target indicates whether it requires a bitcode marker or not.

We'd update the iOS targets (I guess?) to indicate that they need a bitcode marker.

Does that make sense?

(FWIW I also bring this up because the actual code for "embed something empty" vs "embed the bitcode" is really small, it's just an `Option` argument vs not)

Reply:
Ok, I have reinstated the marker section. I found a nicer encoding that uses a new, secondary `BitcodeSection` enum. I haven't done anything to treat iOS targets specially, which means I am preserving the existing behaviour of inserting a marker on any target if `-Zembed-bitcode` is used in combination with a low optimization level.